File: spec2code/java-spring/src/main/java/de/codecentric/javaspring/persistence/AuthorService.java
package de.codecentric.javaspring.persistence;

import static java.util.Comparator.comparing;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;
import org.springframework.stereotype.Service;

@Service
public class AuthorService {
    private final Map<UUID, Author> store = new ConcurrentHashMap<>();

    public Author save(Author author) {
        store.put(author.id(), author);
        return author;
    }

    public Author get(UUID id) {
        if (!store.containsKey(id)) {
            throw new NotFoundException("Author", id);
        }
        return store.get(id);
    }

    public void delete(UUID id) {
        if (!store.containsKey(id)) {
            throw new NotFoundException("Author", id);
        }
        store.remove(id);
    }

    public PaginatedResult<Author> list(String search, int page, int perPage) {
        final List<Author> allAuthors = List.copyOf(store.values());
        final Stream<Author> filteredAuthors;
        if (search == null) {
            filteredAuthors = allAuthors.stream();
        } else {
            filteredAuthors = allAuthors.stream()
                    .filter(author -> author.firstName().toLowerCase().startsWith(search.toLowerCase())
                            || author.lastName().toLowerCase().startsWith(search.toLowerCase()));
        }
        final List<Author> sortedAuthors = filteredAuthors.sorted(comparing(Author::lastName)).toList();

        int totalAuthors = sortedAuthors.size();
        int totalPages = (int) Math.ceil((double) totalAuthors / perPage);

        int start = (page - 1) * perPage;
        int end = Math.min(start + perPage, totalAuthors);

        final List<Author> paginatedAuthors =
                sortedAuthors.subList(Math.min(start, totalAuthors), Math.min(end, totalAuthors));

        return new PaginatedResult<>(paginatedAuthors, totalAuthors, totalPages);
    }
}

File: spec2code/kotlin-spring/src/main/kotlin/de/codecentric/kotlinspring/persistence/AuthorService.kt
package de.codecentric.kotlinspring.persistence

import org.springframework.stereotype.Service
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@Service
class AuthorService {
    private val store: MutableMap<UUID, Author> = ConcurrentHashMap()

    fun get(id: UUID) = store.getOrElse(id) { throw NotFoundException("Author with ID $id not found.") }

    fun save(author: Author): Author {
        store[author.id] = author

        return author
    }

    fun delete(id: UUID) = store.remove(id) ?: throw NotFoundException("Author with ID $id not found.")

    fun list(search: String?): List<Author> =
        search?.let {
            store.values
                .toList()
                .filter { it.firstName.startsWith(search) || it.lastName.startsWith(search) }
        } ?: store.values.toList()
}
